### ISSUE TITLE
Import Realm in Keycloak 18.0.0 - failed

### ISSUE THREAD

**Issue by edwint88** [keycloak/keycloak#11664]
### Describe the bug

I’m using a docker image and I uploaded the _realm.json_ in` /opt/keycloak/data/import` to get the nice auto import functionality.
When I add now to start --import-realm I get:

```
2022-04-26 08:28:56,195 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2022-04-26 08:28:56,195 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to import realm: myrealm
2022-04-26 08:28:56,195 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Script upload is disabled
```

which confuses me, because

Script Upload should be removed (why is import using that?) and
doesn’t work.
Without the `--import-realm` all works fine

Reference to documentation: [https://www.keycloak.org/server/importExport](https://www.keycloak.org/server/importExport)

### Version

18.0.0

### Expected behavior

To import the realms automatically from the `data/import` location.

### Actual behavior

Gives an error

### How to Reproduce?

1. create a dockerfile
2. add a realm.json to the path
3. build keycloak in docker file
4. build dockerfile
5. start container

**Comment by edwint88** [comment]
The same error I get also when I try to upload a realm from the interface/admin console:
`Uncaught server error: java.lang.RuntimeException: Script upload is disabled`

**Comment by stefanmaric** [comment]
Linking this discussion I opened a few ago: https://github.com/keycloak/keycloak/discussions/11658

**Comment by edwint88** [comment]
And also here: https://keycloak.discourse.group/t/import-realm-in-keycloak-18-0-0-failed/15166

**Comment by pedroigor** [comment]
@edwint88 Is your realm declaring providers that rely on scripts such as mappers, authenticators, or policies?

If so, the realm configuration is now invalid and you should not be able to import. See the note [here](https://www.keycloak.org/2022/04/keycloak-1800-released.html) about the removal of the `upload-scripts` feature.

**Comment by stefanmaric** [comment]
> @edwint88 Is your realm declaring providers that rely on scripts such as mappers, authenticators, or policies?

I had the same issue and based on your feedback I search and found I had these policies for one of the realm clients:

```json
"policies": [
  {
    "id": "b56eebd7-8e73-4449-b110-30dfdbc77f03",
    "name": "Default Policy",
    "description": "A policy that grants access only for users within this realm",
    "type": "js",
    "logic": "POSITIVE",
    "decisionStrategy": "AFFIRMATIVE",
    "config": {
      "code": "// by default, grants any permission associated with this policy\n$evaluation.grant();\n"
    }
  },
  { ... },
  {
    "id": "1428ae4c-b767-41b9-aaf9-bd8b0d8497e2",
    "name": "Default Permission",
    "description": "A permission that applies to the default resource type",
    "type": "resource",
    "logic": "POSITIVE",
    "decisionStrategy": "UNANIMOUS",
    "config": {
      "defaultResourceType": "urn:api:resources:default",
      "applyPolicies": "[\"Default Policy\"]"
    }
  },
  {...}
]
```

I checked and these were not referenced anywhere else so I proceeded to delete them and it worked.

I guess it was created automatically by a previous version of KC.

For anyone else experiencing the same issue: search for `"type": "js"` to pin-point the root cause in the realm export file.

**Comment by edwint88** [comment]
> @edwint88 Is your realm declaring providers that rely on scripts such as mappers, authenticators, or policies?
> 
> If so, the realm configuration is now invalid and you should not be able to import. See the note [here](https://www.keycloak.org/2022/04/keycloak-1800-released.html) about the removal of the `upload-scripts` feature.

We have an Authenticator, but not a JavaScript one. It's a simple Java SPI that we add through .jar in `/opt/keycloak/providers` and then, yes, configured in the realm.json. 
Is this affected too?

**Comment by ericwout-overheid** [comment]
We would like to reopen this discussion. We are just starting with a project using Keycloak (18.0.2). We have create a realm via the user interface with some basic settings. We exported the realm using the Manage > Export > Partial Export feature in the user interface.

When we try to import this export using the --import-realm option in a clean container (also 18.0.2), we get the mentioned error. It seems like a bug when Keycloak does not generate a compatible export for the same version?

It works when we remove the policies element enterily (not when we only remove the policy with the "js"-type). But as new users we are not sure whether that's a safe thing to do.

**Comment by edwint88** [comment]
Hi,

Deleting the default policies is safe. Later on you want to have your own
there. The error I think is still relate to the js thing, because if you
delete just that policy some reference will remain to it and therefore will
fail your import.

Cheers

**Comment by ericwout-overheid** [comment]
Ah, thanks for clearing that up. But shouldn't it be registered as a bug that when you create an export, it cannot be imported without removing this by hand (after googling for it?).

**Comment by edwint88** [comment]
Sure, I don't know if there is a ticket already or to be fixed in the next version.

**Comment by yangboyd** [comment]
Change the default policy type to the script-jsfilename:

          {
            "id": "834d0278-8eda-4533-911b-78a1d983ffc9",
            "name": "Default Policy",
            "description": "A policy that grants access only for users within this realm",
            "type": "script-default-policy.js",
            "logic": "POSITIVE",
            "decisionStrategy": "AFFIRMATIVE"
          },


put the jar to providers, then build and start

**Comment by qhaas** [comment]
> shouldn't it be registered as a bug that when you create an export, it cannot be imported without removing this by hand

I agree, this issue should be re-opened, this problem still exists in keycloak 20.0

**Comment by herbertscruz** [comment]
Replacing the `Default Policy` setting with the one below resolves the issue and maintains the expected behavior.

```json
          {
            "id": "98bb844e-5524-4c42-914b-cc5121d3124d",
            "name": "Default Policy",
            "description": "A policy that grants access only for users within this realm",
            "type": "role",
            "logic": "POSITIVE",
            "decisionStrategy": "AFFIRMATIVE",
            "config": {
              "roles": "[{\"id\":\"default-roles-main\",\"required\":false}]"
            }
          }
```

**Comment by ajaiswal595** [comment]
> script-default-policy.js

@yangboyd this works 

put the jar to providers, then
kc.sh build
and
kc.sh start
.
The default-policy.js in one sample jar photoz-uma-js-policies.jar.
[photoz-uma-js-policies.jar](https://github.com/yangboyd/keycloak-quickstarts/blob/a7826b98aab11000901c449b3330b681151f1e81/app-authz-uma-photoz/photoz-uma-js-policies.jar)

But the problem is with upgrade from wildfly(15) distro to keycloak(20) , as in database have of type js entry for default policy. 
<img width="1679" alt="image" src="https://user-images.githubusercontent.com/8745632/209076415-4cb05da3-5d59-4870-af68-d8cab103d8d0.png">


And we can set end user/admin to change the type of Policy type.

What is the recommendation in the scenarios of migration?

**Comment by pedroigor** [comment]
@edwint88 @yangboyd I'm working on it. Can we discuss this issue at https://github.com/keycloak/keycloak/issues/16074?

**Comment by hantsy** [comment]
@qhaas 's solution resolved my problem.

**Comment by goekhantopcu** [comment]
The solution of @qhaas worked for me as well, thanks a lot men <3 (using keycloak 24.0.4)

**Comment by qhaas** [comment]
IMO, given this is still an issue with 25.0.x, likely deserves a new issue.

I'd be interested in someone with more keycloak expertise than I to explain the consequences of [stefanmaric's workaround](https://github.com/keycloak/keycloak/issues/11664#issuecomment-1111062102) vs [herbertscruz's workaround](https://github.com/keycloak/keycloak/issues/11664#issuecomment-1326652539) given defaults / behaviors have likely changed since they were first proposed as a workaround.

**Comment by bornav** [comment]
Why is this closed?
export/import is unsuccessful if docs are followed do to 2 policies that are generated not by the user

**Comment by shawkins** [comment]
@bornav the script upload error was most recently resolved as #32269 - if that is your underlying problem, please see that issue. If you are experiencing something else, then please open a new issue.
